sklearn/utils/graph_shortest_path.pyx

```diff
@@ -1,238 +1,1118 @@
 """
 Routines for performing shortest-path graph searches
 
-The main interface is in the function `graph_shortest_path`.  This
-calls cython routines that compute the shortest path using either
-the Floyd-Warshall algorithm, or Dykstra's algorithm with Fibonacci Heaps.
+The main interface is in the function :func:`shortest_path`.  This
+calls cython routines that compute the shortest path using
+the Floyd-Warshall algorithm, Dijkstra's algorithm with Fibonacci Heaps,
+the Bellman-Ford algorithm, or Johnson's Algorithm.
+
+This module is a backdoor for scipy.sparse.csgraph.
 """
 
 # Author: Jake Vanderplas  -- <vanderplas@astro.washington.edu>
-# License: BSD 3 clause, (C) 2011
+# License: BSD, (C) 2011
+import warnings
 
 import numpy as np
 cimport numpy as np
 
-from scipy.sparse import csr_matrix, isspmatrix, isspmatrix_csr
+from scipy.sparse import csr_matrix, isspmatrix, isspmatrix_csr, isspmatrix_csc
+from .sparsetools._graph_validation import validate_graph
 
 cimport cython
 
 from libc.stdlib cimport malloc, free
 
-np.import_array()
-
 DTYPE = np.float64
 ctypedef np.float64_t DTYPE_t
 
 ITYPE = np.int32
 ctypedef np.int32_t ITYPE_t
 
+# Fused type for int32 and int64
+ctypedef fused int32_or_int64:
+    np.int32_t
+    np.int64_t
+
+# EPS is the precision of DTYPE
+cdef DTYPE_t DTYPE_EPS = 1E-15
+
+# NULL_IDX is the index used in predecessor matrices to store a non-path
+cdef ITYPE_t NULL_IDX = -9999
+
+
+class NegativeCycleError(Exception):
+    def __init__(self, message=''):
+        Exception.__init__(self, message)
 
-def graph_shortest_path(dist_matrix, directed=True, method='auto'):
+
+def shortest_path(csgraph, method='auto',
+                  directed=True,
+                  indices=None,
+                  return_predecessors=False,
+                  unweighted=False,
+                  overwrite=False):
     """
+    shortest_path(csgraph, method='auto', directed=True, return_predecessors=False,
+                  unweighted=False, overwrite=False)
+
     Perform a shortest-path graph search on a positive directed or
     undirected graph.
 
+    .. versionadded:: 0.11.0
+
```

Comment:
This versionadded refers to scipy; it should be taken out here.

Also perhaps add a note that this function is backported from scipy.